src/midi/midi_alsa.cpp

```diff
@@ -25,19 +25,46 @@
 
 #include <string>
 #include <sstream>
+#include <functional>
 
+#include "logging.h"
+#include "programs.h"
 #include "string_utils.h"
 
 #define ADDR_DELIM ".:"
 
-#if (SND_LIB_MAJOR >= 1) || ((SND_LIB_MAJOR == 0) && (SND_LIB_MINOR >= 6))
-#define snd_seq_flush_output(x)           snd_seq_drain_output(x)
-#define snd_seq_set_client_group(x, name) /* nop */
-#define my_snd_seq_open(seqp)             snd_seq_open(seqp, "hw", SND_SEQ_OPEN_OUTPUT, 0)
-#else
-/* SND_SEQ_OPEN_OUT causes oops on early version of ALSA */
-#define my_snd_seq_open(seqp) snd_seq_open(seqp, SND_SEQ_OPEN)
-#endif
+using port_action_t = std::function<void(snd_seq_client_info_t *client_info,
+                                         snd_seq_port_info_t *port_info)>;
+
+void for_each_alsa_seq_port(port_action_t action)
+{
+	// We can't reuse the sequencer from midi handler, as the function might
+	// be called before that sequencer is created.
+	snd_seq_t *seq;
+	if (snd_seq_open(&seq, "hw", SND_SEQ_OPEN_OUTPUT, 0) != 0) {
+		LOG_MSG("ALSA: Error: Can't open MIDI sequencer");
+		return;
+	}
+
+	snd_seq_client_info_t *client_info;
+	snd_seq_client_info_malloc(&client_info);
+
+	snd_seq_port_info_t *port_info;
+	snd_seq_port_info_malloc(&port_info);
+
```

Comment:
If we don't check the rcodes from the two mallocs, we could initialize them as `nullptr` and assert they're populated before using them further below?